### PR TITLE
Adds more tests around the macros for trait edge cases and others

### PR DESCRIPTION
**Pull request by ianoc-stripe**
This should fail in CI, will follow up with code changes to remove warnings and fix these tests.

### CODE REVIEW

File: scalding-serialization/src/main/scala/com/twitter/scalding/serialization/macros/impl/BinaryOrdering.scala [hunk 1]
```diff
@@ -4,10 +4,8 @@ import com.twitter.scalding.serialization.OrderedSerialization
...
-/**
- * @author Mansur Ashraf.
- */
-object BinaryOrdering {
-
+trait BinaryOrdering {
   implicit def ordSer[T]: OrderedSerialization[T] = macro com.twitter.scalding.serialization.macros.impl.OrderedSerializationProviderImpl[T]
 }
+
+object BinaryOrdering extends BinaryOrdering {}
```
Comment (johnynek) [inline scalding-serialization/src/main/scala/com/twitter/scalding/serialization/macros/impl/BinaryOrdering.scala]: I don't *think* you need `{}`

**Comment by ianoc** [comment]
Does a bunch of damage on #1721 but there's more warnings, some of which the linter fires in generated code, not too sure how to stop that even....

@johnynek As you requested :) 


(Also included is failing test + bug fix for some trait handling for classes with lots of fields failing to deserialize. And some better escape hatch handling so users can supply their own OrderedSerializations when the macros fail more often).

File: scalding-serialization/src/main/scala/com/twitter/scalding/serialization/macros/impl/ordered_serialization/providers/ProductOrderedBuf.scala [hunk 2]
```diff
@@ -42,54 +48,182 @@ object ProductOrderedBuf {
...
+    def freshT(id: String) = TermName(c.freshName(id))
 
     val dispatcher = buildDispatcher
     val elementData: List[(c.universe.Type, TermName, TreeOrderedBuf[c.type])] =
-      outerType
-        .declarations
+      outerType.declarations
         .collect { case m: MethodSymbol => m }
-        .filter(m => m.name.toString.startsWith("_"))
+        .filter(m => m.name.toTermName.toString.startsWith("_"))
```
Comment (johnynek) [inline scalding-serialization/src/main/scala/com/twitter/scalding/serialization/macros/impl/ordered_serialization/providers/ProductOrderedBuf.scala]: I had to remove a bunch of these since they were calling `toTermName` on things that were already `TermName`, but I guess with `blackbox` this changed?

**Review by johnynek — APPROVED** [review]
Thanks for removing the curlies!

File: scalding-serialization/src/main/scala/com/twitter/scalding/serialization/macros/impl/ordered_serialization/providers/ProductOrderedBuf.scala (hunk 2 above)
Comment (ianoc-stripe) [reply scalding-serialization/src/main/scala/com/twitter/scalding/serialization/macros/impl/ordered_serialization/providers/ProductOrderedBuf.scala]: I suspect i've glossed over some of those warnings migrating in bulk the code with the fixes in it. I'll do a clean compile or two to make sure i've not added new warnings.

**Comment by ianoc-stripe** [comment]
Ok i think that should be better minus those few fixes.

There are a bunch of warnings about unreachable code in match statements, though they seem spurrious. I've added a sys.error into one of them and it fires...

**Comment by johnynek** [comment]
I'm merging as is and we can deal with the blackbox issue in a follow up (I would love to get the warnings cleaned up in scalding).
